File: articles/behavioral-economics-sustainable-consumption/fortran/adoption_kernel.f90
program adoption_kernel
  implicit none

  real(8) :: income, environmental_concern, present_bias, loss_aversion
  real(8) :: norm_sensitivity, friction_sensitivity, quality_uncertainty
  real(8) :: infrastructure_access, subsidy, norm_signal, friction
  real(8) :: effective_premium, affordability, immediate_cost, utility_diff, probability
  integer :: default_green

  income = 65000.0d0
  environmental_concern = 0.62d0
  present_bias = 0.28d0
  loss_aversion = 2.0d0
  norm_sensitivity = 0.55d0
  friction_sensitivity = 0.50d0
  quality_uncertainty = 0.25d0
  infrastructure_access = 0.60d0
  subsidy = 0.05d0
  default_green = 1
  norm_signal = 0.70d0
  friction = 0.08d0

  effective_premium = max(0.10d0 - subsidy, 0.0d0)
  affordability = 1.0d0 / log(income)
  immediate_cost = effective_premium * affordability * 100.0d0 + friction * friction_sensitivity

  utility_diff = -0.65d0 &
      + 1.10d0 * environmental_concern &
      + 0.72d0 * default_green &
      + 0.85d0 * norm_sensitivity * norm_signal &
      + 0.55d0 * infrastructure_access &
      - 1.75d0 * immediate_cost &
      - 0.38d0 * present_bias &
      - 0.35d0 * loss_aversion * effective_premium &
      - 0.62d0 * quality_uncertainty

  probability = 1.0d0 / (1.0d0 + exp(-utility_diff))

  print *, probability
end program adoption_kernel

File: articles/behavioral-economics-sustainable-consumption/cpp/adoption_kernel.cpp
#include <cmath>
#include <iostream>
#include <vector>

struct Agent {
    double income;
    double environmental_concern;
    double present_bias;
    double loss_aversion;
    double norm_sensitivity;
    double friction_sensitivity;
    double quality_uncertainty;
    double infrastructure_access;
};

double adoption_probability(
    const Agent& a,
    double subsidy,
    int default_green,
    double norm_signal,
    double friction
) {
    double effective_premium = std::max(0.10 - subsidy, 0.0);
    double affordability = 1.0 / std::log(a.income);

    double immediate_cost = effective_premium * affordability * 100.0
        + friction * a.friction_sensitivity;

    double utility_diff =
        -0.65
        + 1.10 * a.environmental_concern
        + 0.72 * default_green
        + 0.85 * a.norm_sensitivity * norm_signal
        + 0.55 * a.infrastructure_access
        - 1.75 * immediate_cost
        - 0.38 * a.present_bias
        - 0.35 * a.loss_aversion * effective_premium
        - 0.62 * a.quality_uncertainty;

    return 1.0 / (1.0 + std::exp(-utility_diff));
}

int main() {
    std::vector<Agent> agents = {
        {45000, 0.55, 0.35, 2.1, 0.50, 0.60, 0.35, 0.40},
        {90000, 0.75, 0.20, 1.8, 0.65, 0.40, 0.20, 0.80}
    };

    for (const auto& a : agents) {
        std::cout << adoption_probability(a, 0.05, 1, 0.70, 0.08) << "\n";
    }

    return 0;
}
